posthog/hogql/printer.py

```diff
@@ -137,6 +137,14 @@ def visit_select_query(self, node: ast.SelectQuery):
             if node.limit_with_ties:
                 clauses.append("WITH TIES")
 
+        if len(node.union_all or []) > 0:
+            # :TRICKY: union queries are not "children" the parent select, but "siblings". Adjust the stack for the visit
```

Comment:
Perhaps `UNION` should be a level above `SELECT`? That's how ClickHouse does this (https://github.com/ClickHouse/ClickHouse/blob/6d48bb25ff2519b84306a7abb1c17ffe60034f20/src/Analyzer/UnionNode.h) – there's `UnionNode` with the constituent queries as `children`. Then on the query tree level, they treat `SELECT ...` and `SELECT ... UNION/EXCEPT/INTERSECT SELECT ...` as interchangeable constructs (https://github.com/ClickHouse/ClickHouse/blob/6d48bb25ff2519b84306a7abb1c17ffe60034f20/src/Analyzer/QueryTreeBuilder.cpp#LL142C33-L142C33).

Reply:
I guess that makes sense. I've updated the code to do just that. Ready for another look.